drivers/audio/dmic_nrfx_pdm.c

```diff
@@ -59,6 +62,14 @@ static void event_handler(const struct device *dev, const nrfx_pdm_evt_t *evt)
 			LOG_ERR("Failed to allocate buffer: %d", ret);
 			stop = true;
 		} else {
+			ret = dmm_buffer_in_prepare(drv_cfg->mem_reg, &buffer,
+						    drv_data->block_size,
+						    (void *)drv_data->mem_slab->buffer);
```

Comment:
I think this is wrong and won't work if non-DMAble buffer is provided - API of `dma_buffer_in_prepare` looks like this: https://github.com/zephyrproject-rtos/zephyr/blob/065bd322200ccf2f9d717fd258ec966a522ce733/soc/nordic/common/dmm.h#L142

So you should have additional pointer `void *dma_buffer` , use it in `dma_buffer_in_prepare` and then in `nrfx_pdm_buffer_set` :
```
struct dmic_nrfx_pdm_drv_data {
	const nrfx_pdm_t *pdm;
	struct onoff_manager *clk_mgr;
        void *memslab_buffer;
};

void event_handler() {
void *dma_buffer;

k_mem_slab_alloc(drv_data->mem_slab, &dev->data->memslab_buffer, K_NO_WAIT);
dmm_buffer_in_prepare(drv_cfg->mem_reg, dev->data->memslab_buffer,
	              drv_data->block_size,
                      &dma_buffer);
nrfx_pdm_buffer_set(drv_data->pdm, dma_buffer, drv_data->block_size / 2);
}
```

Then, when PDM is stopped or buffer is filled with data, and the DMA buffer is to be freed:
```
else if (evt->buffer_released) {
	       ret = dmm_buffer_in_release(drv_cfg->mem_reg, dev->data->memslab_buffer,
						  drv_data->block_size,
						  evt->buffer_released);
		ret = k_msgq_put(&drv_data->rx_queue,
				 dev->data->memslab_buffer,
				 K_NO_WAIT);
               /* User calls k_mem_slab_free on original `dev->data->memslab_buffer`  */
```
Similar code needs to be applied to `	if (drv_data->stopping) {` conditional

It works right now because `dmm` calls does nothing if memslab is completely preallocated in correct region, and this seem to be the case for existing tests